### PR TITLE
Fixed #27767 -- Added distinct argument to ArrayAgg.

### PR DESCRIPTION
**Pull request by orf**
https://code.djangoproject.com/ticket/27767#ticket

### CODE REVIEW

File: tests/postgres_tests/test_aggregates.py [hunk 1]
```diff
@@ -128,6 +128,22 @@ def test_json_agg_empty(self):
         self.assertEqual(values, json.loads('{"jsonagg": []}'))
 
 
+class TestAggregateDistinct(PostgreSQLTestCase):
```
Comment (charettes) [inline tests/postgres_tests/test_aggregates.py]: `TestArrayAggDistinct`?

File: django/contrib/postgres/aggregates/general.py [hunk 2]
```diff
@@ -8,6 +8,10 @@
 
 class ArrayAgg(Aggregate):
     function = 'ARRAY_AGG'
+    template = '%(function)s(%(distinct)s%(expressions)s)'
+
+    def __init__(self, *args, distinct=False, **kwargs):
+        super(ArrayAgg, self).__init__(*args, **kwargs, distinct='DISTINCT ' if distinct else '')
```
Comment (charettes) [inline django/contrib/postgres/aggregates/general.py]: Use `super()`, master is Python 3 only.

Comment (orf) [inline django/contrib/postgres/aggregates/general.py]: Done :)

File: tests/postgres_tests/test_aggregates.py (hunk 1 above)
Comment (orf) [inline tests/postgres_tests/test_aggregates.py]: Done :)

**Comment by charettes** [comment]
This will also require docs. I think the `DISTINCT` subclass you've proposed on the ticket makes sense.

**Comment by orf** [comment]
Is it OK to do the subclass as a separate PR? I'll do some docs changes now

**Comment by charettes** [comment]
> Is it OK to do the subclass as a separate PR?

This could a separate commit in this PR.

e.g. a first one that adds the subclass and a second one titled "Fixed #27767 -- Added support DISTINCT in ArrayAgg." that does everything related to `ARRAY_AGG(DISTINCT)`.

**Comment by orf** [comment]
I've made some documentation changes, let me know if they are OK. It could be a separate commit, but I'd rather add a separate ticket and do it there as it might require more changes than I'm comfortable with at this moment. I'm still somewhat new to the internals of Django!

**Comment by charettes** [comment]
> I've made some documentation changes, let me know if they are OK. It could be a separate commit, but I'd rather add a separate ticket and do it there as it might require more changes than I'm comfortable with at this moment. I'm still somewhat new to the internals of Django!

Sure!

**Comment by orf** [comment]
Done @charettes, let me know if it's all OK

File: docs/ref/contrib/postgres/aggregates.txt [hunk 3]
```diff
@@ -22,9 +22,16 @@ General-purpose aggregation functions
...
 ------------
 
-.. class:: ArrayAgg(expression, **extra)
+.. class:: ArrayAgg(expression, distinct=False, **extra)
 
     Returns a list of values, including nulls, concatenated into an array.
+    
+    .. attribute:: distinct
+
+        .. versionadded:: 2
```
Comment (timgraham) [inline docs/ref/contrib/postgres/aggregates.txt]: 2.0
Please also add release notes in `docs/releases/2.0.txt`.

File: tests/postgres_tests/test_aggregates.py [hunk 4]
```diff
@@ -128,6 +128,24 @@ def test_json_agg_empty(self):
...
+    @classmethod
+    def setUpTestData(cls):
+        AggregateTestModel.objects.create(char_field='Foo')
+        AggregateTestModel.objects.create(char_field='Foo')
+        AggregateTestModel.objects.create(char_field='Bar')
+
+    def test_string_agg_distinct_false(self):
+        values = AggregateTestModel.objects.aggregate(arrayagg=ArrayAgg('char_field', distinct=False))
+        agg = list(sorted(values['arrayagg']))
+        self.assertEqual(agg, ['Bar', 'Foo', 'Foo'])
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: I don't think `list()` is needed. `self.assertEqual(sorted(values['arrayagg'], ['Bar', 'Foo', 'Foo'])` looks good to me.

File: tests/postgres_tests/test_aggregates.py [hunk 5]
```diff
@@ -128,6 +128,24 @@ def test_json_agg_empty(self):
...
 
 
+class TestArrayAggDistinct(PostgreSQLTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        AggregateTestModel.objects.create(char_field='Foo')
+        AggregateTestModel.objects.create(char_field='Foo')
+        AggregateTestModel.objects.create(char_field='Bar')
+
+    def test_string_agg_distinct_false(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: I think the `TestStringAggregateDistinct` class could be reused considering the setup methods are the same -- just remove `String` from the class name.

File: tests/postgres_tests/test_aggregates.py (hunk 4 above)
Comment (orf) [inline tests/postgres_tests/test_aggregates.py]: Done 👍

File: tests/postgres_tests/test_aggregates.py (hunk 5 above)
Comment (orf) [inline tests/postgres_tests/test_aggregates.py]: Good idea, done

File: tests/postgres_tests/test_aggregates.py [hunk 6]
```diff
@@ -146,6 +146,16 @@ def test_string_agg_distinct_true(self):
         self.assertEqual(values['stringagg'].count('Bar'), 1)
 
 
+class TestArrayAggregateDistinct(TestStringAggregateDistinct):
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: Using a subclass isn't correct. That will cause the tests in the super class to be executed again.

File: docs/releases/2.0.txt [hunk 7]
```diff
@@ -72,6 +72,9 @@ Minor features
 :mod:`django.contrib.postgres`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+* A `distinct` keyword has been added to `ArrayAgg` to specify if array values
```
Comment (timgraham) [inline docs/releases/2.0.txt]: Follow the style of the 1.11 release notes: "The new distinct argument for ArrayAgg determines if aggregated values will be distinct." The backticks need to be doubled to have an effect on the rendered text.

File: tests/postgres_tests/test_aggregates.py (hunk 6 above)
Comment (orf) [inline tests/postgres_tests/test_aggregates.py]: Oh right, sorry, I just understood your previous comment on this file! I will fix them now. I'm using the github editor to do this, and there is no way to squash commits (apart from when merging). I can do it locally when I have access to my machine later if it's an issue.

**Comment by timgraham** [comment]
merged in b5393028bfc939adf14d0fa5e4088cddd3b9dfa1, thanks!

**Pull request closed without merging**